mmdet/datasets/pipelines/auto_augment.py

```diff
@@ -0,0 +1,142 @@
+import copy
+
+import numpy as np
+
+from mmdet.core import BitmapMasks
+from ..builder import PIPELINES
+
+
+@PIPELINES.register_module()
+class AutoAugment(object):
+    """Auto augmentation.
+
+    This data augmentation is proposed in
+    "Learning Data Augmentation Strategies for Object Detection".
+    Link: https://arxiv.org/pdf/1906.11172.pdf
+    Required key is "img", and optional keys are "gt_bboxes", "gt_masks",
+    "gt_seg".
+
+    Args:
+        auto_augment_policies (list[list[dict]]): The policies of auto
+            augmentation. Each element in auto_augment_policies is a
+            specifically augmentation policy, and is composed by several
+            augmentations (dict). When call AutoAugment, a random element in
+            auto_augment_policies will be selected to augment the image.
+            Defaults: None.
+
+    Examples:
+        >>> replace = (104, 116, 124)
+        >>> auto_augment_policies = [
+        >>>     [
+        >>>         dict(type='Sharpness', prob=0.0, level=8),
+        >>>         dict(
+        >>>             type='Shear',
+        >>>             prob=0.4,
+        >>>             level=0,
+        >>>             replace=replace,
+        >>>             axis='x')
+        >>>     ],
+        >>>     [
+        >>>         dict(
+        >>>             type='Rotate',
+        >>>             prob=0.6,
+        >>>             level=10,
+        >>>             replace=replace),
+        >>>         dict(type='Color', prob=1.0, level=6)
+        >>>     ]
+        >>> ]
+        >>> augmentation = AutoAugment(auto_augment_policies)
+        >>> img = np.ones(100, 100, 3)
+        >>> gt_bboxes = np.ones(10, 4)
+        >>> results = dict(img=img, gt_bboxes=gt_bboxes)
+        >>> results = augmentation(results)
+    """
+
+    def __init__(self, auto_augment_policies=None):
+        assert type(auto_augment_policies) is list and len(
+            auto_augment_policies) > 0, \
+            'The type and length of auto_augment_policies' \
+            'must be list and more than 0, respectively.'
+        for one_augment_policy in auto_augment_policies:
+            assert type(one_augment_policy) is list and len(
+                one_augment_policy) > 0, \
+                'The type and length of each element in' \
+                'auto_augment_policies must be list and more than 0, ' \
+                'respectively.'
+            for augment in one_augment_policy:
+                assert type(augment) is dict and 'type' in augment, \
+                    'The type of each specifically augmentation must be dict' \
+                    "and contain the key of 'type'."
+
+        self.auto_augment_policies = auto_augment_policies.copy()
+        raise NotImplementedError('Auto augmentation is working in progress '
+                                  'and currently not callable.')
```

Comment:
This error can be removed. As we did not provide configs before it is finished, most users will not use it.